x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/aggs/frequentitemsets/FrequentItemSetsAggregationBuilder.java

```diff
@@ -0,0 +1,168 @@
+        if (minimumSupport <= 0 || minimumSupport > 1) {
+            throw new IllegalArgumentException(
+                "[minimum_support] must be greater than 0 and less or equal to 1. Found [" + minimumSupport + "] in [" + name + "]"
+            );
+        }
+        this.minimumSupport = minimumSupport;
+        if (minimumSetSize <= 0) {
+            throw new IllegalArgumentException(
+                "[minimum_set_size] must be greater than 0. Found [" + minimumSetSize + "] in [" + name + "]"
+            );
+        }
+        this.minimumSetSize = minimumSetSize;
+        if (size <= 0) {
+            throw new IllegalArgumentException("[size] must be greater than 0. Found [" + size + "] in [" + name + "]");
+        }
+        this.size = size;
+    }
+
+    public FrequentItemSetsAggregationBuilder(StreamInput in) throws IOException {
+        super(in);
+        this.fields = in.readList(MultiValuesSourceFieldConfig::new);
+        this.minimumSupport = in.readDouble();
+        this.minimumSetSize = in.readInt();
+        this.size = in.readInt();
+    }
+
+    @Override
+    protected AggregationBuilder shallowCopy(Builder factoriesBuilder, Map<String, Object> metadata) {
+        return new FrequentItemSetsAggregationBuilder(name, fields, minimumSupport, minimumSetSize, size);
+    }
+
+    @Override
+    public String getType() {
+        return NAME;
+    }
+
+    @Override
+    protected void doWriteTo(StreamOutput out) throws IOException {
+        out.writeList(fields);
+        out.writeDouble(minimumSupport);
+        out.writeInt(minimumSetSize);
+        out.writeInt(size);
+    }
+
+    @Override
+    protected AggregatorFactory doBuild(AggregationContext context, AggregatorFactory parent, Builder subfactoriesBuilder)
+        throws IOException {
+
+        return new FrequentItemSetsAggregatorFactory(
+            name,
+            context,
+            parent,
+            subfactoriesBuilder,
+            metadata,
+            fields,
+            minimumSupport,
+            minimumSetSize,
+            size
+        );
+    }
+
+    @Override
+    protected XContentBuilder internalXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject();
+        builder.startArray(FIELDS.getPreferredName());
+        for (MultiValuesSourceFieldConfig field : fields) {
+            field.toXContent(builder, params);
+        }
+        builder.endArray();
+        builder.field(MINIMUM_SUPPORT.getPreferredName(), minimumSupport);
+        builder.field(MINIMUM_SET_SIZE.getPreferredName(), minimumSetSize);
+        builder.field(Aggregation.CommonFields.SIZE.getPreferredName(), size);
+        builder.endObject();
+        return builder;
+    }
+
+    @Override
+    public BucketCardinality bucketCardinality() {
+        return BucketCardinality.NONE;
+    }
+
+    @Override
+    public Version getMinimalSupportedVersion() {
+        return Version.V_8_3_0;
+    }
+
```

Comment:
```suggestion
        @Override
        public final AggregationBuilder subAggregations(Builder subFactories) {
            throw new AggregationInitializationException(
                "Aggregator [" + name + "] of type [" + getType() + "] cannot accept sub-aggregations"
            );
        }
```

Or something as this doesn't handle sub aggs yet.

Reply:
I have to check this, there is something called `LeafOnly` which I should use, I don't recall why I couldn't use it.

Reply:
alright and for the record: I can't use the other abstraction as e.g. `MultiValuesSourceAggregationBuilder` does not support multiple types.